packages/vitest-pool-workers/src/pool/index.ts

```diff
@@ -649,15 +651,37 @@ function maybeGetResolvedMainPath(project: Project): string | undefined {
 	}
 }
 
-async function runTests(
+type RunTestsArgs = [
 	ctx: Vitest,
 	mf: Miniflare,
 	workerName: string,
 	project: Project,
 	config: ResolvedConfig,
 	files: string[],
-	invalidates: string[] = []
-) {
+	invalidates?: string[],
+];
+
+/**
+ * Ensures that test runs are queued if all available threads are in use.
+ */
+function queuedTestRun() {
+	const threadPool = new ResourcePool(calculateAvailableThreads());
+
+	return async function run(...args: RunTestsArgs): Promise<void> {
+		await threadPool.nextAvailableResource();
+
+		// Promise.finally will be called in the event of either resolving or rejecting,
+		// so the thread will always be released unless the test run hangs.
+		return runTests(...args).finally(() => {
+			threadPool.releaseResource();
+		});
+	};
```

Comment:
Could we lock this section of code using a form such as:

```
const limiter = new Semaphore(calculateAvailableThreads);
limiter.runWith(() => runTests(...args));
});
```

where the `runWith()` function will handle locking, waiting and unlocking the semaphore?

Reply:
I like that - it's a simpler API.